File: objectiveai/objectiveai/src/chat/completions/http.rs
use crate::{HttpClient, HttpError};
use futures::Stream;

pub async fn create_chat_completion_unary(
    client: &HttpClient,
    mut params: super::request::ChatCompletionCreateParams,
) -> Result<super::response::unary::ChatCompletion, HttpError> {
    params.stream = None;
    client
        .send_unary(reqwest::Method::POST, "chat/completions", Some(params))
        .await
}

pub async fn create_chat_completion_streaming(
    client: &HttpClient,
    mut params: super::request::ChatCompletionCreateParams,
) -> Result<
    impl Stream<
        Item = Result<
            super::response::streaming::ChatCompletionChunk,
            HttpError,
        >,
    >,
    HttpError,
> {
    params.stream = Some(true);
    client
        .send_streaming(reqwest::Method::POST, "chat/completions", Some(params))
        .await
}

File: objectiveai-js/src/chat/completions/http.ts
import { ObjectiveAI, RequestOptions } from "../../client";
import { Stream } from "../../stream";
import { ChatCompletion } from "./response/unary/chatCompletion";
import {
  ChatCompletionCreateParams,
  ChatCompletionCreateParamsStreaming,
  ChatCompletionCreateParamsNonStreaming,
} from "./request/chatCompletionCreateParams";
import { ChatCompletionChunk } from "./response/streaming/chatCompletionChunk";

export function create(
  client: ObjectiveAI,
  body: ChatCompletionCreateParamsStreaming,
  options?: RequestOptions,
): Promise<Stream<ChatCompletionChunk>>;
export function create(
  client: ObjectiveAI,
  body: ChatCompletionCreateParamsNonStreaming,
  options?: RequestOptions,
): Promise<ChatCompletion>;
export function create(
  client: ObjectiveAI,
  body: ChatCompletionCreateParams,
  options?: RequestOptions,
): Promise<Stream<ChatCompletionChunk> | ChatCompletion> {
  if (body.stream) {
    return client.post_streaming<ChatCompletionChunk>(
      "/chat/completions",
      body,
      options,
    );
  }
  return client.post_unary<ChatCompletion>("/chat/completions", body, options);
}
